feat: Add popup events to GA4

leogermani opened the pull request:
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request, and the reason for such changes. -->

Connects the Popup events to GA4.

Note that the Popup events have already been tested. This only connects them to GA4. It also adds a little more detail to the newsletter subscription events.

This PR also fixes a bug in the `donation_new` event and the `form_submission_success` event for donations. Both when using Stripe as platform.

Since the new donation in Stripe is triggered when Stripe server hits our webhook, the GA Client and Session IDs were not bein fetched by the cookie. So I had to add these information in the payment metadata and retrieve them for the order when they are marked as completed.

### How to test the changes in this Pull Request:

Since we changed the structure of the data and also fixed a bug, we should do an extensive tests of as many events as possible.

Get the GA4 credentials. You will need your measurement ID and an API secret.

* api_secret - Required. An API SECRET generated in the Google Analytics UI. To create a new secret, navigate to:
    Admin > Data Streams > choose your stream > Measurement Protocol > Create
* measurement_id - Required. The measurement ID associated with a stream. Found in the Google Analytics UI under:
    Admin > Data Streams > choose your stream > Measurement ID

Store this info in the database:
```
wp option set ga4_measurement_id "G-XXXXXXXXXX"
wp option set ga4_api_secret YYYYYYYYYYYYYYYYYYY
```
Now, let's test:

1. Make sure RAS is active and that you have a registered reader
2. Make sure to add this constant to wp-config
```
define( 'NEWSPACK_LOG_LEVEL', 2 );
define( 'NEWSPACK_EXPERIMENTAL_GA4_EVENTS', true );
```
4. Checkout the newsletters plugin on https://github.com/Automattic/newspack-newsletters/pull/1111
5. Setup a popup with a Newsletter subscription form
6. Subscribe to a newsletter via this popup
8. Visit the Google Analytics Realtime dashboard
9. See the `campaign_interaction` events coming in

![Captura de tela de 2023-03-09 13-15-59](https://user-images.githubusercontent.com/971483/224086546-fc9ee3ac-a5e7-4a08-9921-7c00d78b5ed5.png)

11. Inspect them and see check its attributes to see if the values are correct.

![Captura de tela de 2023-03-09 13-51-25](https://user-images.githubusercontent.com/971483/224097513-bf0700aa-c552-4e66-bc8e-358fd1c08279.png)

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

<!-- Mark completed items with an [x] -->

miguelpeixe commented on includes/data-events/class-popups.php:
```diff
@@ -180,18 +180,21 @@ public static function registration_submission_with_status( $email, $user_id, $p
 			[
-				'action' => self::FORM_SUBMISSION,
+				'action'                         => self::FORM_SUBMISSION,
+				'newsletter_subscription_method' => $metadata['newsletters_subscription_method'],
```
For the sake of preserving the root object of `campaign_interaction` more agnostic and with permanent keys, I'd place this type of metadata nested in an isolated key since it's regarding a block inside the campaign. Something like `campaign_data` or `blocks_data`?

Something similar is done here, for donations:

https://github.com/Automattic/newspack-plugin/blob/aa5543a7c507b673650b0bb9ce9af0fd4a3965d3/includes/data-events/listeners.php#L219-L233

The platform may vary and have specific metadata associated with it. That shouldn't mess with the root object.

miguelpeixe replied:
Also, documenting in `README.md` also helps with perspective on how we're building these.

leogermani replied:
> Something similar is done here, for donations:

IN donations all info about the donation is block specific as well. `amount`, `currency`, etc... It's the same case. If I change here I'd have to change there as well.

I get the point, can I just ask to do this refactor on another PR? because this will also touch the gha-4 connector and there's another PR open and they'll all conflict if I start this now.

> Also, documenting in README.md also helps with perspective on how we're building these.

I'm doing this! :) Since I'm touching this in several PRs and they are still changing a bit I'll do it all at once as one PR at the end.

GA4 connector will have its own README as well.

miguelpeixe replied:
> IN donations all info about the donation is block specific as well. amount, currency, etc... It's the same case. If I change here I'd have to change there as well.

I don't understand the relation. Every `donation_new` must have `amount`, `currency`, etc... It's a requirement of the action event regardless of the platform.

The `campaign_interaction` have some clear requirements, like `action` and `referer`. Other arguments are specific to the contents of the campaign, therefore not always there. If the event was called `campaign_interaction_donation` or `campaign_interaction_newsletter_subscription` it would be a different case.

We're not ensuring a good schema here.

miguelpeixe replied:
These listeners are not just to serve GA4, they are general purpose.

leogermani replied:
ah, the other PR got merged. I can do this refactor now.

I can add `registration_data`, `donation_data` and `newsletter_subscription_data` keys in the root and place the values there. WDYT?

miguelpeixe replied:
`interaction_data` sounds like a good candidate as well for holding these

miguelpeixe replied:
why not an abstract one, so APIs don't have to struggle with it? if the action tells what it is, why would they key also have to?

leogermani replied:
> interaction_data sounds like a good candidate as well for holding these

To ensure we always have the same schema. And to be consistent with `has_registration`, `has_newsletter` and `has_donation`.

I know it's unlikely, but in theory you can have more than one of these blocks inside a popup.

Also, we can know beforehand the schema inside each of these keys as well.

>  if the action tells what it is, why would the key also have to?

The action does not tell us that. At least not yet. that's being done by the `has_*` params

miguelpeixe commented:
This also goes for the `campgaign_has_{block-name}_block`. If we manage to take that information out of the schema and in a `has_blocks` array containing the names of the block, it's also a more scalable and generic solution for campaign prompts.

leogermani commented:
> This also goes for the campgaign_has_{block-name}_block. If we manage to take that information out of the schema and in a has_blocks array containing the names of the block, it's also a more scalable and generic solution for campaign prompts.

agreed

> My suggestion is we use the action value to inform what type of interaction it is, and an abstract key like interaction_data to hold every interaction metadata.

This keeps the schema simpler, on one hand. But on the other hand I have no idea what's inside `interaction_data`. If we have one for each type, we can also describe what they look like.

It's not like we are going to have dozens of interaction types there.

miguelpeixe commented:
> But on the other hand I have no idea what's inside interaction_data. If we have one for each type, we can also describe what they look like.

Looks to me like an easier documentation problem to tackle, not a schema one.

leogermani commented:
> Looks to me like an easier documentation problem to tackle, not a schema one.

![Captura de tela de 2023-03-10 10-47-58](https://user-images.githubusercontent.com/971483/224332324-248f1b24-238d-492b-93b3-cf4858bf0841.png)
(you win)

miguelpeixe commented:
That must be a first! 😆

leogermani commented:
Just found a bug that I don't know how we didn't catch before. Will address it

leogermani commented:
Still working on this big gap I've found on tracking Stripe donation events in GA4. will need more time on Monday. I'll mark the PR as needs review again once it's done

leogermani commented:
@miguelpeixe 

This is ready for review again. I also fixed a bug (described it in the PR description).

I also added the documentation.

Since we touched the structure of all events, I think we should do an extensive tests of as many events as possible, as well as donations (one time and recurring) with both platforms.

leogermani commented:
@miguelpeixe :wave:

miguelpeixe commented:
Sorry for the delay, I plan on reviewing this today.

miguelpeixe commented on includes/data-events/connectors/ga4/class-ga4.php:
```diff
@@ -259,6 +305,31 @@ public static function handle_newsletter_subscribed( $params, $data ) {
+		unset( $transformed_data['ga_params'] );
+		unset( $transformed_data['ga_client_id'] );
+
+		unset( $transformed_data['interaction_data'] );
+		$transformed_data = array_merge( $transformed_data, $data['interaction_data'] );
+
+		unset( $transformed_data['campaign_blocks'] );
```
When the `reader_registered` and `newsletter_subscribed` events are sent, they are populated with these campaign metadata but not sanitized like here for `campaign_interaction`.

Should some handling happen for such cases or just let the generic sanitizer handle it? This is how it's currently handled:

```
[NEWSPACK-DATA-EVENTS-GA4][Newspack\Data_Events\Connectors\GA4::log]: Parameter campaign_blocks has an invalid value. It was removed from the np_newsletter_subscribed event.
[NEWSPACK-DATA-EVENTS-GA4][Newspack\Data_Events\Connectors\GA4::log]: Parameter interaction_data has an invalid value. It was removed from the np_newsletter_subscribed event.
```

leogermani replied:
weird, I didn't see this happening. I'm removing these 2 keys from the data and they should never be sent in tha GA event...

I'll double check

`campaign_blocks` is transformed into the `has_XX_block` params

`interaction_data` is merged and moved to the same level as root params...

miguelpeixe replied:
They are but only in the context of the `campaign_interaction` event, not the other ones that also end up getting populated with the same values.

leogermani replied:
https://github.com/Automattic/newspack-plugin/pull/2337/commits/4fa17fa9cd685c9cd3042e36664a0c609c9e9c24

miguelpeixe commented on includes/data-events/connectors/ga4/class-ga4.php:
```diff
@@ -139,6 +147,44 @@ public static function filter_event_body( $body, $event_name ) {
 	}
 
+	/**
+	 * When the donation_new event is dispatched from the Stripe webhook, it can't catch GA client ID from Cookies.
```
This is also an issue with other logic inside `GA4::filter_event_body()`. The event triggered by Stripe webhook is being dispatched with `logged_in` set to `no` and without `is_reader`, even though I'm a logged-in reader.

leogermani replied:
I was tired when I noticed this the other day.

Looking at this again today I was happy to see it was an easy fix after all I did.

I also figured we could default the value of `is_reader` to `no` if the user is not logged in

https://github.com/Automattic/newspack-plugin/pull/2337/commits/df620d276bb0a0417399b94fb4b997f1e6c3f576

miguelpeixe commented on includes/data-events/README.md:
```diff
@@ -139,16 +177,43 @@ When a WooCommerce Subscription status changes.
+| `action_type`        | `string` | `donation`, `registration` or `newsletters_subscription`                                                                                       |
+| `interaction_data`   | `array`  | Depending on the actino type, it will return different information about the interaction.                                                      |
+
+#### structure of interaction_data:
```
```suggestion
#### Possible values for `interaction_data`
```

miguelpeixe reviewed (changes requested):
Thank you for the documentation!

kmwilkerson commented:
@yogeshbeniwal That's correct. Due to GA4's [automatically collected events](https://support.google.com/analytics/answer/9234069?hl=en), NCI (who produces the News Tagging Guide) will be updating the guide to use those auto events so that users no longer have to manually create events.
